added the hinge_embedding_loss

sushmanthreddy opened the pull request:
close #21076

sushmanthreddy commented:
@MahmoudAshraf97 can u pls review th pulll request pls??

sushmanthreddy commented:
@MahmoudAshraf97 @YushaArif99 can u pls review my pull request, i have added the support for hinge_embeddding_loss ,and tests are passing in local system and the sc is attached over here..pls review and let me knew any changes are needed.

sushmanthreddy commented:
checked supporting datatypes by this method
![Screenshot from 2023-08-20 02-30-25](https://github.com/unifyai/ivy/assets/73489688/8db229fc-25bc-49a3-997e-7115bbe532cf)

MahmoudAshraf97 commented on ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py:
```diff
@@ -19,7 +19,7 @@
     ),
     dtype_and_log_input=helpers.dtype_and_values(
-        available_dtypes=helpers.get_dtypes("float"),
+        available_dtypes=helpers.get_dtypes("int"),
```
revert this change

MahmoudAshraf97 commented on ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py:
```diff
@@ -57,3 +57,52 @@ def test_log_poisson_loss(
+        min_value=-3,
+        max_value=3,
+        allow_inf=False,
+        min_num_dims=1,
+        max_num_dims=3,
+        min_dim_size=3,
```
remove these unnecessary constraints, the test should be as unconstrained as possible
```suggestion

```

MahmoudAshraf97 commented on ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py:
```diff
@@ -57,3 +57,52 @@ def test_log_poisson_loss(
+        min_value=-1,
+        max_value=1,
+        allow_inf=False,
+        min_num_dims=1,
+        max_num_dims=3,
+        min_dim_size=3,
```
```suggestion

```

MahmoudAshraf97 commented on ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py:
```diff
@@ -57,3 +57,52 @@ def test_log_poisson_loss(
+
+
+# hinge_embedding_loss
+@handle_test(
+    fn_tree="functional.ivy.experimental.hinge_embedding_loss",
+    dtype_and_input=helpers.dtype_and_values(
+        available_dtypes=helpers.get_dtypes("float"),
+        min_value=-3,
+        max_value=3,
+        allow_inf=False,
+        min_num_dims=1,
+        max_num_dims=3,
+        min_dim_size=3,
+    ),
+    dtype_and_target=helpers.dtype_and_values(
+        available_dtypes=helpers.get_dtypes("float"),
+        min_value=-1,
+        max_value=1,
+        allow_inf=False,
+        min_num_dims=1,
+        max_num_dims=3,
+        min_dim_size=3,
+    ),
```
you can also generate two arrays with the same strategy instead of having two strategies

MahmoudAshraf97 commented on ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py:
```diff
@@ -57,3 +57,52 @@ def test_log_poisson_loss(
+        min_dim_size=3,
+    ),
+    margin=st.floats(min_value=0, max_value=3, allow_infinity=False),
+    test_with_out=st.just(False),
```
why is this hardcoded to false?
```suggestion
```

MahmoudAshraf97 commented on ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py:
please address the concerns in the testing review, also add testing for reduction argument

MahmoudAshraf97 reviewed (changes requested):
Hello @sushmanthreddy , the function argument should be unified across all backends and still satisfy the superset behaviour, please address the issues mentioned in the review

sushmanthreddy commented:
@MahmoudAshraf97 I have implemented the changes u have mentioned.And ,I have  added the reduction parameter too in test file.It just got out of sight.
And after reading superset behaviour ,have understood the backend implementation too ,and removed unecessary parameters which are not present in other backends.
One strategy have true labels and other have targets which are binary  1 and -1 ,so it better to have two different strategies ,now tests are passing.can u pls review it ,if changes are needed pls let me knew otherwise pls merge it

sushmanthreddy commented:
![Screenshot from 2023-08-20 20-28-08](https://github.com/unifyai/ivy/assets/73489688/b151d9ee-06ed-49d9-afe4-686ab41a4aca)
tests passing screenshot

MahmoudAshraf97 commented on ivy/functional/backends/paddle/experimental/losses.py:
```diff
@@ -0,0 +1,43 @@
+        target,
+        margin=margin,
+        reduction=reduction,
+        out=out,
```
paddle function doesn't have an out argument

MahmoudAshraf97 commented on ivy/functional/backends/torch/experimental/losses.py:
```diff
@@ -0,0 +1,38 @@
+        target,
+        margin=margin,
+        reduction=reduction,
+        out=out,
```
torch function doesn't have an out argument

MahmoudAshraf97 commented on ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py:
```diff
@@ -57,3 +57,49 @@ def test_log_poisson_loss(
+        target=target_data[0],
+        margin=margin,
+        reduction=reduction,
+        atol_=1e-2,
+        rtol_=1e-2,
```
this is a relatively high tolerance please remove it

MahmoudAshraf97 reviewed (changes requested):
Hi, please remove all tolerances not just relative
the out argument doesn't exist in the frameworks but it should stay in the function signature, please check how `handle_out_argument` works and make sure it wraps the main function

YushaArif99 commented:
Hi, @sushmanthreddy. I'll be closing this PR has it has been inactive for over a month now. Thanks! 🙂
